### PR TITLE
Prepare for using the initial solution from ebos directly 

### PR DESCRIPTION
**Pull request by totto82**
- Let ebos handle the swatinit scaling
- Keep the initial solution a bit longer so flow can output it. 
- Stop using the equilGrid in the init code. 
(The equilGrid can most probably be removed now. But not in this PR.) 

Note. Merge together with associated PR in opm-simulators.

### CODE REVIEW

**Comment by andlaus** [comment]
this looks good and makes things simpler, so I'm all for merging. Nevertheless, I've got three minor questions/demands:

* can you please squash these commits together? I'd like to keep the git history for the master branch clean, i.e., debug commits  should not be in it IMO
* can this be merged without preconditions or does it need patches for upstream modules?
* does this affect the results or the performance of flow in any significant way?

**Comment by andlaus** [comment]
The second point is already answered by your PR description: OPM/opm-simulators#1355 is needed. I somehow missed this, sorry.

**Comment by totto82** [comment]
> can you please squash these commits together? I'd like to keep the git history for the master branch clean, i.e., debug commits should not be in it IMO

That was my plan. I wasn't 100% sure about the last commit so I kept the option to revert it easily. 
Can EquilGrid be completely removed now? I have tested that this patch works with MPI. 

> can this be merged without preconditions or does it need patches for upstream modules?

NO

> does this affect the results or the performance of flow in any significant way?

I guess the time for the initialization is halved, but other than that NO.

**Comment by andlaus** [comment]
> Can EquilGrid be completely removed now?

if the hydrostatic code from opm-core now works with a distributed grid, yes. does it?

File: ebos/eclproblem.hh [hunk 1]
```diff
@@ -556,6 +551,10 @@ public:
             // write the summary information after each time step
             summaryWriter_.write(wellManager_);
         }
+
+        // we no longer need the initial soluiton
+        if (this->simulator().episodeIndex() == 0)
+            initialFluidStates_.clear();
```
Comment (andlaus) [inline ebos/eclproblem.hh]: why did you move this from `initialSolutionApplied()` to here?

Comment (totto82) [reply ebos/eclproblem.hh]: Its needed for output in flow (check the associated PR in opm-simulators).

**Comment by totto82** [comment]
> if the hydrostatic code from opm-core now works with a distributed grid, yes. does it?

Yes it does. Could you check if it work with Alugrid as well? Or is this tested automatically.

File: ebos/eclproblem.hh (hunk 1 above)
Comment (andlaus) [reply ebos/eclproblem.hh]: hm, I had a brief look and it looks like one can just use the fluid state from the `IntensiveQuantities` there? the solution should not have changed compared to the initial one, so if there is a difference in the initial fluid state and `iq.fluidState()` that would be a bug as far as I can see.

**Comment by andlaus** [comment]
> Yes it does.

cool, then all these `equilGrid` hacks can be removed!

> Could you check if it work with Alugrid as well? Or is this tested automatically.

It's not tested automatically. I'll give it a shot and report back.

**Comment by andlaus** [comment]
Okay, I've tested: It currently does not compile with alugrid. That's because `Opm::initStateEquil()` uses `Opm::UgGridHelpers` extensively and this is neither available nor easy to provide for AluGrid. As far as I can see, it is not very difficult to replace these constructs by DUNE equivalents in the `Opm::initStateEquil()` code, but that would kill compatibility with the old `UnstructuredGrid` approach. (Not sure if this is acceptable at this stage.)

It also did not compile with `PolyhedralGrid`, but this seemed to be for different reasons related to some issues within `PolyhedralGrid`.

**Comment by totto82** [comment]
OK. I will revert the removal of the EquilGrid in this PR since it is not the main target of this PR.

**Comment by andlaus** [comment]
> OK. I will revert the removal of the EquilGrid in this PR since it is not the main target of this PR.

as you prefer. I think that the issue with the alternative grids should not block the EquilGrid removal, at least not if the init code is scheduled to be adapted to properly work with Dune grids in the not-too-distant future...

File: ebos/eclproblem.hh (hunk 1 above)
Comment (totto82) [reply ebos/eclproblem.hh]: You are almost right. The state (rs, rv) is changed for cells with only water.

Comment (andlaus) [reply ebos/eclproblem.hh]: > The state (rs, rv) is changed for cells with only water.

I see. since Rs and Rv does not matter in this case, wouldn't it be better to output the initial solution which is actually used by the simulator? (granted, that would almost certainly require another update of the reference data for the unit tests.)

Comment (totto82) [reply ebos/eclproblem.hh]: Yes, it doesn't matter for the results, but our customer wanted it this way. I think the solution I suggest here is as clean as it gets without moving the output code to ewoms.

**Comment by totto82** [comment]
> as you prefer. I think that the issue with the alternative grids should not block the EquilGrid removal, at least not if the init code is scheduled to be adapted to properly work with Dune grids in the not-too-distant future...

I will go for the current solution then. ie no changes to current PR. 
It seems the polyhedral grid is not working for flow with the current master anyway. And ebos itself seem to compile  and run fine.

File: ebos/eclproblem.hh (hunk 1 above)
Comment (andlaus) [reply ebos/eclproblem.hh]: yeah, it's not a big deal. the question I'm more concerned about is: isn't what we currently do misleading?

Also, I've just looked a bit closer: is `initialFluidStates_` even accessible from outside of the problem class? I did not see any accessor method...

File: ebos/eclproblem.hh [hunk 2]
```diff
@@ -1026,48 +1007,9 @@ public:
...
-            priVars0 = priVars1;
-        }
-
-        // the solution (most likely) has changed, so we need to invalidate the cache for
-        // intensive quantities
-        this->simulator().model().invalidateIntensiveQuantitiesCache(/*timeIdx=*/0);
+    // temporary solution to facilitate output of initial state from flow
+    const ScalarFluidState& initialFluidState(unsigned globalDofIdx ) const {
+        return initialFluidStates_[globalDofIdx];
     }
```
Comment (totto82) [inline ebos/eclproblem.hh]: @andlaus Here it is.

File: ebos/eclproblem.hh (hunk 1 above)
Comment (andlaus) [reply ebos/eclproblem.hh]: forget the second comment: I was on master instead of on your branch...

File: ebos/eclproblem.hh (hunk 2 above)
Comment (andlaus) [reply ebos/eclproblem.hh]: yes. That was a case of "lost in branches" for me.
